### PR TITLE
version: add an AtLeast helper to compare versions.

### PR DESCRIPTION
**Pull request by danderson**
Signed-off-by: David Anderson <danderson@tailscale.com>

---
This change is [Reviewable](https://reviewable.io/reviews/tailscale/tailscale/458)

### CODE REVIEW

File: version/cmp.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+// negatives, for example when an OSS build supports a given feature,
+// but AtLeast is called with an official release number as the
+// minimum
+//
+// version and minimum can both be either an official Tailscale
+// version numbers (major.minor.patch-extracommits-extrastring), or an
+// OSS build datestamp (date.YYYYMMDD). For Tailscale version numbers,
+// AtLeast also accepts a prefix of a full version, in which case all
+// missing fields are assumed to be zero.
+func AtLeast(version string, minimum string) bool {
```
Comment (bradfitz) [inline version/cmp.go:31]: Being in this package, I assume this compared against the current version.

Maybe add `func Cmp(a, b string) int` and then make AtLeast only take a String?

File: version/cmp.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	case v.Patch != m.Patch:
+		return v.Patch > v.Patch
+	default:
+		return v.ExtraCommits > m.ExtraCommits
+	}
+}
+
+type parsed struct {
+	Major, Minor, Patch, ExtraCommits int // for Tailscale version
+	Datestamp                         int // for OSS version
```
Comment (bradfitz) [inline version/cmp.go:68]: give example values in quotes in comment?

File: version/cmp.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+		return v.Datestamp >= m.Datestamp
+	case v.Major == m.Major && v.Minor == m.Minor && v.Patch == m.Patch && v.ExtraCommits == m.ExtraCommits:
+		// Exactly equal Tailscale versions
+		return true
+	case v.Major != m.Major:
+		return v.Major > m.Major
+	case v.Minor != m.Minor:
+		return v.Minor > m.Minor
+	case v.Patch != m.Patch:
+		return v.Patch > v.Patch
```
Comment (bradfitz) [inline version/cmp.go:60]: real

File: version/cmp.go (hunk 1 above)
Comment (bradfitz) [reply version/cmp.go:31]: Also fine as-is. Maybe renamed? Reads a bit odd as-is.

Comment (danderson) [reply version/cmp.go:31]: So, I started out with a Cmp, but the problem is that our version numbers don't have a single total order. There is no return value for Cmp("date.1234", "0.98") that makes sense.

Instead, I tried to make a function that is scoped to specifically "can you tell me for sure that version is > minimum? It's okay to say no if unsure." Such a function can't be used to sort an array, only to do stuff like version gating (which is what I need this for).

Re: package: I need this in the control server to compare client-reported versions, so I can't make the comparison implicit against the burned-in version. But it's still utilities for manipulating version numbers, so this feels like it should be the place for it?

Dunno. I can also make it into a corp-only helper package, if you prefer that.

File: version/cmp.go (hunk 3 above)
Comment (danderson) [reply version/cmp.go]: Oops, sigh.

File: version/cmp.go (hunk 1 above)
Comment (danderson) [reply version/cmp.go:31]: Going to submit as-is so I can work on the control server, but happy to figure out a better place/name for this. I agree that a binary comparator is a bit weird within this package.

**Pull request merged**